### ISSUE TITLE
Seams appear in images with transparency when rotated

### ISSUE THREAD

**Issue by pearcetm** [openseadragon/openseadragon#2294]
Seams appearing between tiles has been a topic of frequent discussion, including a relatively recent enhancement (https://github.com/openseadragon/openseadragon/pull/2075) that has solved many problems (e.g. https://github.com/openseadragon/openseadragon/discussions/2085).

I'm having trouble with seams appearing specifically when images are rotated to non-cardinal orientations (anything other than 0, 90, 180, 270 degrees). The property `subPixelRoundingForTransparency` does not seem to solve the problem for rotated images that have transparency.

To see the problem in action, the [ui-rotation demo page](https://openseadragon.github.io/examples/ui-rotation/) can be modified in developer mode. In the console, paste the following: `viewer.world.getItemAt(0).source.hasTransparency=()=>true;` and use the slider on the bottom viewer to adjust the rotation angle. Dark seams appear between tiles (since the background is dark). (Note: This may or may not happen on non-retina screens, I haven't tested it yet on non-HD screens.)

I may have missed something - is there a different setting I should try out? Any thoughts on how to improve stitching of transparent images of arbitrary rotation angles would be much appreciated!

**Comment by iangilman** [comment]
Good catch! Thank you for the easy repro steps! For reference, here's what it looks like: 

<img width="802" alt="image" src="https://user-images.githubusercontent.com/338989/217954228-3feaa863-b184-4c2b-8a6d-3714e3d93e92.png">

Unfortunately, as far as I'm aware, this is essentially a browser bug. Possible workarounds I can think of: 

* In this case, expand the tiles just a little (when drawing) so they overlap each other just a little. Of course, if any of the semi-transparent areas of the tiles are along tile edges, you might get seams as they double up. Also, this would distort the image just a little bit.  
* This is a giant project, but I wonder if a WebGL renderer would be able to avoid these issues. 
* Figure out some way to fill in areas along seams with bits from the adjacent tiles. That sounds like direct pixel manipulation, though, and would likely not perform well. 

I think that first option could potentially work for some scenarios, and maybe it could be tune-able so it could work in a lot of scenarios with a little bit of tweaking. Making something that works in all scenarios might be challenging. It's all about trade-offs.

**Comment by pearcetm** [comment]
@iangilman Thanks for taking a look. I haven't had a chance yet to do a deep dive on how the tiles are drawn onto the canvas, and how the `subPixelRoundingForTransparency` interacts with this. If it is indeed a browser bug, it's pretty consistent across Chrome and Firefox (mac) as well as Chrome on windows, and is present on retina and standard def displays.

Regarding the first option you suggest, I've played around with increasing the tile overlap setting, to no avail - it does not seem to impact the seams in the slightest. Would "expanding the tiles a little" as you suggest need to be implemented in core OSD rather than by the overlap?

Regarding the second option (WebGL), I too was wondering if this may be a solution. Many years ago I did a bit of work with OpenGL via OpenSceneGraph, and the viewport transforms (including rotation) were essentially independent of how the underlying scene was composed, so I could see WebGL being similar and not susceptible to the same issues that canvas implementations have. @Aiosa I know you've done some work with WebGL rendering in OSD - any thoughts on how hard it would be to compose the tiled image view via WebGL rather than canvas?

I agree that option 3 (pixel manipulation on every draw) is probably not the way to go... 

Given the consistency of this across at least a subset of browsers, I still wonder whether there may be something about the OSD implementation rather than a cross-browser canvas bug. Something to explore further, perhaps...

**Comment by Aiosa** [comment]
For a WebGL renderer, first step would be to remove all rendering logics from `Tile` and move it to the drawer class. Then, one could - like tile sources - implement their way of drawing. The actual drawing process would be pretty straightforward and simple, you would have to just implement a correct vertex shader (my guess is ~ 10/15 lines of glsl code). I could even re-use some of the utilities from my WebGL lib, for instance. Some plugins could then integrate more complex renderers such as deck.gl (I would not put this to the core though). The rendering issues would IMHO disappear.

 More problematic would be dealing with textures and other features that are otherwise simple with canvas, such as rendering debug info. You would need a generic interface of loading data to the GPU if it comes from cache, adding additional functions to the `TileCache` (now it has image and canvas getters) would not scale well. Such interface would have to handle both WebGl 1 and 2; not too hard but also not so simple. I guess that could be the drawer responsibility, if it got the image data. The debug info could be rendered as another image and then sent within a texture, slow but simplest solution, it's debugging so that would be enough. 

In summary, the actual rendering would be rather simple to implement (if you stick with 2D) and actually allow much more features other than rotation such as scaling, distortion maps etc... (just a hinting the WebGL renderer should be extensible),  but fine-tuning would take some time.

**Comment by Aiosa** [comment]
I will probably try implementing it after I deal with more urgent things, in 5 months or so.

**Comment by iangilman** [comment]
@Aiosa Awesome! I agree with your assessment, and I'd love to see what you come up with whenever you can get to it :) 

@pearcetm My understanding of the issue is that the anti-aliasing at the edge of one tile plus the anti-aliasing at the edge of its neighbor tile doesn't add up to fully opaque. I think this may be fundamental to the canvas anti-aliasing technique. You can certainly do some experiments by drawing two images at an angle next to each other in a canvas (without using OSD). If you can come up with a technique there that works, we can translate it to OSD. 

As for expanding the tiles, I was thinking that would happen dynamically while drawing. Maybe that could happen in `Tile.drawCanvas`. 

BTW, I think the main discussion behind seams is https://github.com/openseadragon/openseadragon/issues/1683... Linking it here for reference.

**Comment by pearcetm** [comment]
@iangilman I can confirm it is definitely canvas-related, and there are subtle differences across browsers.

Here's a super basic test:
``` 
      var canvas = document.getElementById('test-canvas');// width == height == 400
      var context = canvas.getContext("2d"); 
      context.fillRect(0, 0, 400, 400); // fill entire canvas with black rectangle
      context.translate(200, 200); // translate to center
      context.rotate(1 * Math.PI/180); // rotate by 1 degree
      context.clearRect(-100, -100, 200, 200); // clear a 200 x 200 rectangle
      context.fillRect(-100, -100, 200, 200); // draw a 200 x 200 rectangle in exactly the same place
```

In Chrome on a Macbook:
<img width="412" alt="image" src="https://user-images.githubusercontent.com/12499997/218277915-00f39d21-e538-41a6-a5de-b34fb5ae08d6.png">

In Firefox and Safari on a Macbook:
<img width="410" alt="image" src="https://user-images.githubusercontent.com/12499997/218277942-3ec65831-b2ad-48ac-a013-08719f3585ad.png">

Zooming into the bottom left corner of each shows the differences:
<img width="197" alt="image" src="https://user-images.githubusercontent.com/12499997/218277990-24aa78f3-6415-4974-8868-059da594a0fe.png">
<img width="234" alt="image" src="https://user-images.githubusercontent.com/12499997/218278018-ed4adf72-768c-429f-9de6-34de6f9b0345.png">

So, one major issues is that there's a mismatch between `clearRect` and `drawRect`/`drawImage` - even when these are the same size, the antialiasing operations are not inverses of each other.

However, it's not just about `clearRect` either. Here's a slight modification that draws two rectangles next to each other to fill in the square:
```
        context.clearRect(-100, -100, 200, 200); // clear a 200 x 200 rectangle
        context.fillRect(-100, -100, 100, 200); // draw a 100 x 200 rectangle in the left half
        context.fillRect(-00, -100, 100, 200); // draw a 100 x 200 rectangle in the right half
```
and the result in Chrome (firefox is similar in appearance):
<img width="226" alt="image" src="https://user-images.githubusercontent.com/12499997/218278601-c215d762-98f8-4655-a930-0a590fd56cb7.png">

So I guess it does seem that the antialiasing done by `canvas` is the fundamental problem. I'll look into potential workarounds, but ultimately I think webgl may really be the better approach.

**Comment by iangilman** [comment]
@pearcetm Perfect... This is exactly the sort of test I was hoping for! Of course it's disappointing that it's in canvas, but it's what I expected. Good to have it laid out this simply. 

Presumably it can be fixed by turning off anti-aliasing (looks like it can be done for image drawing at least with `context.imageSmoothingEnabled`), but of course then we have rough edges. I suppose we could draw the interior parts of the tile without anti-aliasing and the exterior parts with (using multiple draw calls and clipping or something). Of course, that would have a performance effect. I guess it depends on how far we want to go.

**Comment by iangilman** [comment]
Right on. There's been a lot of discussion about making a WebGL drawer throughout the years... Have you seen any of those issues? #68 is the original one and there's also #1482 (although I think the latter may be more complicated than necessary, and includes some things that have already been implemented). Anyway, it would be fantastic to get it rolling if you're up for it!

**Comment by pearcetm** [comment]
As noted in the previous discussion about adding WebGL functionality, there is quite a bit of complexity. This will be a major enhancement.

In order to avoid needing to write WebGL directly (as a first step at least), I've been putting together a renderer that uses [three.js](https://github.com/mrdoob/three.js/) to handle the scene composition and viewport manipulations. I will link a PR that shows the new renderer in a demo page in the next few days. So far it incorporates most functionality of the existing viewer (that I'm aware of, at least). I haven't started looking into how to implement clipping and overlaying the debugging grid quite yet. I can confirm it solves the bug with seams appearing on rotated images with transparency. It should also enable some nice 3D transformations to be added in the future...

Since introducing external libraries as dependencies (or bundling them) isn't really how I think we'd want to proceed, I've taken the approach of implementing the renderer as a plugin. To support this, OSD core needs to be modified a bit, with a couple extra API functions, events, and bug fixes to allow rendering plugins to get all information needed for rendering through event handlers and API calls. Finally, if a renderer is passed in, the built-in drawer needs to be skipped (I haven't implemented this yet, but it shouldn't be too hard). This is somewhat less of a major overhaul than @Aiosa described above, but it is working nicely so far.

Assuming this works, I think it would be feasible eventually to use the overall strategy laid out in the plugin code to implement a pure WebGL version (i.e. avoiding three.js dependency) and integrate that as a second drawer in the OSD core, which could then be automatically used when WebGL support is detected.

@iangilman @Aiosa  does this seem like a reasonable approach?

**Comment by Aiosa** [comment]
Well, most of the proposed changes from my side relate to refactoring, since there is plenty of code in `Tile` that should be implemented by the drawing strategy instead (`if (!drawCanvas) then...`). So the WebGL renderer would be _the third_ drawer (HTML and canvas are already there). 

>To support this, OSD core needs to be modified a bit, with a couple extra API functions, events, and bug fixes to allow rendering plugins to get all information needed for rendering through event handlers and API calls.

I am not in favor of adding a new API just for this; to me, it seems like a temporary solution I would not like to keep backward compatibility for, instead of fixing the problem at the root. The drawer should, as tile sources do, drive the whole process it is (should be) responsible for, e.g. drawing. The reason it cannot access all the necessary data for drawing is that it is not programmed in a flexible way. I would rather refactor the code than add new events.

Other than that, I fully agree.

**Comment by iangilman** [comment]
@pearcetm Sounds awesome! I look forward to the PR :) I agree with @Aiosa's concerns about opening up new APIs, but as you say, it all depends upon the details.

**Comment by pearcetm** [comment]
Turns out that even to just enable plugins to implement drawing logic required a big overhaul. Moving some drawing logic out of `Tile` as @Aiosa mentioned, as well as out of `TiledImage`, and moving some stuff from `Drawer` into other places as well. I think doing this makes sense from a separation of concerns standpoint almost independent of the new rendering logic I've been working on.

It took a bit longer than I'd hoped, but a work-in-progress PR is now up! https://github.com/openseadragon/openseadragon/pull/2310 - I would love to get your comments and suggestions there @iangilman and @Aiosa.

**Comment by iangilman** [comment]
@pearcetm This is very exciting! ❤️

**Comment by pearcetm** [comment]
> @pearcetm Is this fixed by #2310?

You bet! Closing this issue.
